### PR TITLE
Allow users to enable / disable discovered docker images

### PR DESCRIPTION
**Pull request by mfranciszkiewicz**
- new: `DockerWhitelistRPC` for exposing Whitelist methods via RPC
- emit events when new images are discovered
- remember recently discovered images (including basic stats)

### CODE REVIEW

File: golem/node.py [hunk 1]
```diff
@@ -38,7 +38,8 @@
 from golem.ethereum.transactionsystem import TransactionSystem
 from golem.model import DB_MODELS, db, DB_FIELDS
 from golem.network.transport.tcpnetwork_helpers import SocketAddress
-from golem.report import StatusPublisher, Component, Stage, report_calls
+from golem.report import StatusPublisher, Component, Stage, report_calls, \
+    EventPublisher
```
Comment (Wiezzel) [inline golem/node.py]: I encourage using this style according to our guide
```python
from golem.report import (
    StatusPublisher,
    ...
)
```

File: golem/task/server/whitelist.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    def __init__(self) -> None:
+        self._discovered: List[DiscoveredDockerImage] = list()
+
+    def _docker_image_discovered(
+            self,
+            name: str
+    ) -> None:
+        if Whitelist.is_whitelisted(name):
+            return
+        for discovered in self._discovered:
```
Comment (Wiezzel) [inline golem/task/server/whitelist.py]: This definitely looks like a job for a `dict`. Why iterate over all discovered images if you can just check it in constant time?

File: golem/task/server/whitelist.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            for discovered in self._discovered
+        }
+
+    @staticmethod
+    @rpc_utils.expose('env.docker.images.whitelist')
+    def _docker_whitelist_get() -> List[str]:
+        return Whitelist.get()
+
+    @rpc_utils.expose('env.docker.images.whitelist.add')
+    def _docker_whitelist_add(self, image_name: str) -> None:
```
Comment (Wiezzel) [inline golem/task/server/whitelist.py]: This is potentially misleading. Someone might think it adds only this particular image to the whitelist while it actually adds all images from a repository.

(*Actually, we're using the term repository wrong here, but let's keep this aside*)

File: golem/task/server/whitelist.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        repository = repository_from_image_name(image_name)
+        Whitelist.add(repository)
+        self._docker_refresh_discovered_images()
+
+    @staticmethod
+    @rpc_utils.expose('env.docker.images.whitelist.remove')
+    def _docker_whitelist_remove(image_name: str) -> None:
```
Comment (Wiezzel) [inline golem/task/server/whitelist.py]: See comment for `_docker_whitelist_add`

File: golem/envs/docker/whitelist.py [hunk 5]
```diff
@@ -1,7 +1,21 @@
...
 
 
+def repository_from_image_name(image_name: str) -> str:
+    return image_name.split('/')[0]
+
+
 class Whitelist:
+
+    @staticmethod
+    def get() -> List[str]:
```
Comment (Wiezzel) [inline golem/envs/docker/whitelist.py]: `get_all()` maybe?

**Review by Wiezzel — COMMENTED** [review]
Some more and less important remarks and one general idea:

How about changing `DockerWhitelist` model into something like `DockerWhiteAndBlackList` (needs a better name for sure) with `whitelisted` boolean field? Then no separate storing of observed images would be needed and the whole thing would be persistent.

File: golem/task/server/whitelist.py (hunk 2 above)
Comment (mfranciszkiewicz) [reply golem/task/server/whitelist.py]: To keep them in order - might as well be done by `OrderedDict`

Comment (Wiezzel) [reply golem/task/server/whitelist.py]: AFAIK dicts are insert-ordered in Python 3.6+

Comment (mfranciszkiewicz) [reply golem/task/server/whitelist.py]: That's a broader topic but that statement is true for cpython

**Comment by mfranciszkiewicz** [comment]
@Wiezzel I agree, it sounds like a separate task though. We'd need to update stats for every (including known) image and maintain a separate maximum of N observed images in the same table (or create a new one)

**Review by Wiezzel — APPROVED** [review]

**Pull request merged**